File: JavaPools/PoolsLibrary/src/main/java/ru/urfu/javapools/poolslibrary/objectactions/notification/UserDefinedActionError.java
package ru.urfu.javapools.poolslibrary.objectactions.notification;

public class UserDefinedActionError<TV> {

    private UserDefinedActionType _userDefinedActionType;
    private TV _object;
    private Exception _exception;
    
    @Override
    public boolean equals(Object obj) {
    	
    	if (null == obj)
    		return false;
    	if (this == obj)
    		return true;    	    	
    	if (!(obj instanceof UserDefinedActionError<?>))
    		return false;    	    	

		@SuppressWarnings("rawtypes")
		UserDefinedActionError another = (UserDefinedActionError)obj;

        return _userDefinedActionType == another._userDefinedActionType
               && _object.equals(another._object)
               && _exception.getClass() == another._exception.getClass();
    }

    @Override
    public int hashCode() {
        return _userDefinedActionType.hashCode();
    }

	public UserDefinedActionType getUserDefinedActionType() {
		return _userDefinedActionType;
	}

	public TV getObject() {
		return _object;
	}

	public Exception getException() {
		return _exception;
	}

	public UserDefinedActionError<TV> setUserDefinedActionType(UserDefinedActionType userDefinedActionType) {
		_userDefinedActionType = userDefinedActionType;
		return this;
	}

	public UserDefinedActionError<TV> setObject(TV object) {
		_object = object;
		return this;
	}

	public UserDefinedActionError<TV> setException(Exception exception) {
		_exception = exception;
		return this;
	}
}

File: CSharpPools/PoolsLibrary/ObjectActions/Notification/UserDefinedActionError.cs
using System;

namespace PoolsLibrary.ObjectActions.Notification
{
    public class UserDefinedActionError<TV>
    {
        public UserDefinedActionType UserDefinedActionType { get; set; }
        public TV Object { get; set; }
        public Exception Exception { get; set; }

        public override bool Equals(object obj)
        {
            var anotherData = obj as UserDefinedActionError<TV>;
            if (anotherData == null)
                return false;

            return UserDefinedActionType == anotherData.UserDefinedActionType
                   && Object.Equals(anotherData.Object)
                   && Exception.GetType() == anotherData.Exception.GetType();
        }

        public override int GetHashCode()
        {
            return UserDefinedActionType.GetHashCode();
        }
    }
}